synapse/handlers/room_list.py

```diff
@@ -135,114 +147,151 @@ async def _get_public_room_list(
-                "avatar_url": room["avatar"],
-                "world_readable": room["history_visibility"]
-                == HistoryVisibility.WORLD_READABLE,
-                "guest_can_join": room["guest_access"] == "can_join",
-                "join_rule": room["join_rules"],
-                "room_type": room["room_type"],
-            }
+        for room in local_public_rooms:
+            num_joined_members_buckets.setdefault(room.num_joined_members, []).append(
+                room
+            )
 
-            # Filter out Nones – rather omit the field altogether
-            return {k: v for k, v in entry.items() if v is not None}
+        for module_index, fetch_public_rooms in enumerate(
+            self._module_api_callbacks.fetch_public_rooms_callbacks
+        ):
+            # Ask each module for a list of public rooms given the last_joined_members
+            # value from the since token and the probing limit
+            # last_joined_members needs to be reduce by one if this module has already
+            # given its result for last_joined_members
+            module_last_joined_members = last_joined_members
+            if module_last_joined_members is not None and last_module_index is not None:
+                if module_index < last_module_index:
+                    module_last_joined_members = module_last_joined_members - 1
+
+            module_public_rooms = await fetch_public_rooms(
+                network_tuple,
+                search_filter,
+                probing_limit,
+                (
+                    module_last_joined_members,
+                    last_room_id if last_module_index == module_index else None,
+                ),
+                forwards,
+            )
+
+            for room in module_public_rooms:
+                num_joined_members_buckets.setdefault(
+                    room.num_joined_members, []
+                ).append(room)
+                room_ids_to_module_index[room.room_id] = module_index
+
+        nums_joined_members = list(num_joined_members_buckets.keys())
+        nums_joined_members.sort(reverse=forwards)
 
-        results = [build_room_entry(r) for r in results]
+        results = []
+        for num_joined_members in nums_joined_members:
+            results += num_joined_members_buckets[num_joined_members]
 
         response: JsonDict = {}
         num_results = len(results)
-        if limit is not None:
-            more_to_come = num_results == probing_limit
+        if limit is not None and probing_limit is not None:
+            more_to_come = num_results >= probing_limit
 
-            # Depending on direction we trim either the front or back.
-            if forwards:
-                results = results[:limit]
-            else:
-                results = results[-limit:]
+            results = results[:limit]
         else:
             more_to_come = False
 
+        if not forwards:
+            results.reverse()
```

Comment:
Won't this mean that paginating backwards will return rooms with the same member count in a different order than when paginating forwards? The original approach with slicing either the front or back of the results would preserve the order of rooms within a page.